src/model/Enemy.hs

```diff
@@ -1,10 +1,131 @@
-data FastPlayerSeekingEnemy = FastPlayerSeekingEnemy {fastSeekingPos :: Position}
+-- All five enemy data types, enemy type class.
+data BasicEnemy = BasicEnemy {basicEnemyPos :: Position, basicEnemyCooldown :: Int} 
+data BurstEnemy = BurstEnemy {burstEnemyPos :: Position, burstEnemyCooldown :: Int} 
+data ConeEnemy = ConeEnemy {coneEnemyPos :: Position, coneEnemyCooldown :: Int} 
+data BasicPlayerSeekingEnemy = BasicPlayerSeekingEnemy {basicSeekingPos :: Position, basicSeekingCooldown :: Int} 
+data FastPlayerSeekingEnemy = FastPlayerSeekingEnemy {fastSeekingPos :: Position} 
+
+
+-- All enemies are instances of this class.
+class Enemy a where 
+    -- Given an enemy and a bullet, return these arguments if the enemy is hit for despawning both. 
+    hit:: a -> Bullet -> Maybe (a, Bullet) 
+    -- Render enemy at a random vertical position.
+    spawn :: a
+    -- Return the position of an Enemy
+    pos :: a -> Position
+
+-- Default values, change these later.
+screenWidth :: Int
+screenWidth = 500
+randomY :: Int
+randomY = 300   
+
+-- Implementation of method 'hit' for each given enemy. 
+enemyIsHit :: Enemy e => e -> Bullet -> Maybe (e, Bullet)
+enemyIsHit e b | pos e == bulletPos b = Just (e, b) -- Enemy is hit. Change to this later: if the position is approximately equal.
+               | otherwise = Nothing -- Enemy is not hit.
+
+-- Enemy types as instances of type class Enemy (Too much repitition?)
+instance Enemy BasicEnemy where
+        hit = enemyIsHit
+        spawn = BasicEnemy {basicEnemyPos = (screenWidth, randomY) , basicEnemyCooldown = cooldown} -- Spawn an enemy at right with a random y pos.
+                where -- Default value, change this later.
+                        cooldown = 2
+        pos = basicEnemyPos
+
+instance Enemy BurstEnemy where
+        hit = enemyIsHit
+        spawn = BurstEnemy {burstEnemyPos = (screenWidth, randomY) , burstEnemyCooldown = cooldown}  
+                where
+                        cooldown = 3
+        pos = burstEnemyPos
+
+instance Enemy ConeEnemy where
+        hit = enemyIsHit
+        spawn = ConeEnemy {coneEnemyPos = (screenWidth, randomY) , coneEnemyCooldown = cooldown}    
+                where
+                        cooldown = 2
+        pos = coneEnemyPos
+
+instance Enemy BasicPlayerSeekingEnemy where
+        hit = enemyIsHit
+        spawn = BasicPlayerSeekingEnemy {basicSeekingPos = (screenWidth, randomY) , basicSeekingCooldown = cooldown}      
+                where
+                        cooldown = 1
+        pos = basicSeekingPos
+
+instance Enemy FastPlayerSeekingEnemy where
+        hit = enemyIsHit
+        spawn = FastPlayerSeekingEnemy {fastSeekingPos = (screenWidth, randomY)}     
+        pos = fastSeekingPos
+       
+
+-- All enemies, except the player seeking types, are instances of this class.
+class MoveBasic a where 
+        moveBasic :: a -> a 
+    
+instance MoveBasic BasicEnemy where
+        moveBasic b = let (x, y) = pos b in
```

Comment:
Ik denk dat je voor deze en `moveSeeking` hetzelfde kan doen als voor `hit`, een losse functie maken die de duplicatie weghaalt.

Reply:
Bij moveBasic en movePlayerSeeking loop ik er tegenaan dat de field names, dus basicEnemyPos, burstEnemyPos etc, niet in een losse functie kan zetten.

Reply:
Ahh ja dat is fair. Dan laten we het gewoon zo, het zal ook wel gewoon werken.